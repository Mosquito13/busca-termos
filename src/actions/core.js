import axios from 'axios';

import actionTypes from '../constants/actionTypes';
import releaseUtils from '../utils/releaseUtils';
import coreSelectors from '../selectors/core';
import apiUtils from '../utils/apiUtils';
import url from '../constants/url';

export const setLoading = (loading) => ({
  type: actionTypes.SET_LOADING,
  loading
});

export const setData = (data) => ({
  type: actionTypes.SET_DATA,
  data
});

export const setSelectedItemId = (id) => ({
  type: actionTypes.SET_SELECTED_ITEM_ID,
  id
});

export const setFilter = (filter) => ({
  type: actionTypes.SET_FILTER,
  filter
});

export const setAppVersion = (version) => ({
  type: actionTypes.SET_APP_VERSION,
  version
});

export const setAppHasUpdate = (appHasUpdate, updateURL) => ({
  type: actionTypes.SET_APP_HAS_UPDATE,
  appHasUpdate,
  updateURL
});

export const checkForUpdates =
  (currentVersion) => async (dispatch, getState) => {
    const state = getState();
    const hasUpdate = coreSelectors.getAppHasUpdate(state);

    if (hasUpdate === null) {
      const releases = await axios.get(url.LIST_APP_RELEASES);
      const [latestReleaseVersion, latestReleaseURL] =
        releaseUtils.parseLastRelease(releases);

      const hasUpdate = releaseUtils.hasUpdate(
        currentVersion,
        latestReleaseVersion
      );

      dispatch(setAppHasUpdate(hasUpdate, latestReleaseURL));
    }
  };

export const loadData = (settings) => async (dispatch) => {
  const data = await apiUtils.loadData(settings?.languageFolder);
  const appVersion = await apiUtils.loadAppVersion();

  dispatch(setAppVersion(appVersion));
  dispatch(checkForUpdates(appVersion));
  dispatch(setData(data));
  dispatch(setLoading(false));
};

export const openBrowserWithURL = (url) => async () => {
  apiUtils.openBrowserWithURL(url);
};
