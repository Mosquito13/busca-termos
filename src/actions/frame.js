import actionTypes from '../constants/actionTypes';
import apiUtils from '../utils/apiUtils';

export const setMaximized = (value) => ({
  type: actionTypes.SET_MAXIMIZED,
  value
});

export const close = () => () => apiUtils.close();

export const minimize = () => () => apiUtils.minimize();

export const maximize = () => () => apiUtils.maximize();

export const unmaximize = () => () => apiUtils.unmaximize();

export const registerFrameListeners = () => (dispatch) => {
  apiUtils.registerMaximizeListener(() => {
    dispatch(setMaximized(true));
  });

  apiUtils.registerUnmaximizeListener(() => {
    dispatch(setMaximized(false));
  });
};
