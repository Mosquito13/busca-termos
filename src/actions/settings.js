import apiUtils from '../utils/apiUtils';
import actionTypes from '../constants/actionTypes';
import storageUtils from '../utils/storageUtils';
import coreActions from './core';

const settingsActions = {
  setValid(value) {
    return {
      type: actionTypes.SET_VALID,
      value
    };
  },

  setValidating(value) {
    return {
      type: actionTypes.SET_VALIDATING,
      value
    };
  },

  setLanguageFolder(value) {
    return {
      type: actionTypes.SET_LANGUAGE_FOLDER,
      value
    };
  },

  setLanguageFolderError(value) {
    return {
      type: actionTypes.SET_LANGUAGE_FOLDER_ERROR,
      value
    };
  },

  setMainLanguage(value) {
    return {
      type: actionTypes.SET_MAIN_LANGUAGE,
      value
    };
  },

  toggleTranslation(id, value) {
    return {
      type: actionTypes.TOGGLE_TRANSLATION,
      id,
      value
    };
  },

  toggleCompactLayout(value) {
    return {
      type: actionTypes.TOGGLE_COMPACT_LAYOUT,
      value
    };
  },

  toggleDarkTheme(value) {
    return {
      type: actionTypes.TOGGLE_DARK_THEME,
      value
    };
  },

  setSettings(value) {
    return {
      type: actionTypes.SET_SETTINGS,
      value
    };
  },

  validateAndSaveFirstSettings(languageFolder) {
    return async (dispatch) => {
      dispatch(this.setValidating(true));

      const isValid = await apiUtils.validateLanguageFolder(languageFolder);

      if (!isValid) {
        dispatch(this.setLanguageFolderError('Caminho invÃ¡lido.'));
        dispatch(this.setValid(false));
      } else {
        dispatch(coreActions.setLoading(true));
        dispatch(this.setValid(true));
        dispatch(this.saveSettings({ languageFolder }));
      }

      dispatch(this.setValidating(false));
    };
  },

  loadSettings() {
    return async (dispatch) => {
      const settings = storageUtils.getSettings();
      const isValid = await apiUtils.validateLanguageFolder(settings?.languageFolder);

      if (!isValid) {
        dispatch(this.setValid(false));
        dispatch(coreActions.setLoading(false));
      } else {
        dispatch(this.setSettings(settings));
        dispatch(coreActions.loadData(settings));
      }
    };
  },

  saveSettings(values) {
    return {
      type: actionTypes.SAVE,
      values
    };
  }
};

export default settingsActions;
