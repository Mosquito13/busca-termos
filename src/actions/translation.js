import actionTypes from '../constants/actionTypes';
import coreSelectors from '../selectors/core';

export const setTranslation = (id, value) => ({
  type: actionTypes.SET_TRANSLATION,
  id,
  value
});

export const loadTranslation = (id, termId) => (dispatch, getState) => {
  const state = getState();
  const data = coreSelectors.getData(state);
  const translation = data[id].find((d) => d.id === termId);

  dispatch(setTranslation(id, translation?.content));
};
