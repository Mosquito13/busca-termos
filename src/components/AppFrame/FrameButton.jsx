import classNames from 'classnames';

const FrameButton = ({ close, icon, onClick }) => {
  const classes = classNames(
    'app-frame__button',
    close && 'app-frame__button--close'
  );

  return (
    <button className={classes} onClick={onClick}>
      <div className="app-frame__button-icon">
        {icon}
      </div>
    </button>
  );
};

export default FrameButton;
