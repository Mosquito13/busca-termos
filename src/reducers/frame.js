import actionTypes from '../constants/actionTypes';

const initial = {
  maximized: false
};

const frameReducer = (state = initial, action = {}) => {
  switch (action.type) {
    case actionTypes.SET_MAXIMIZED:
      return { ...state, maximized: action.value };

    default:
      return state;
  }
};

export default frameReducer;
