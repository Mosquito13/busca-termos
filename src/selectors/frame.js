const getSlice = (state) => state.Frame;

const frameSelectors = {
  isMaximized(state) {
    return getSlice(state).maximized;
  }
};

export default frameSelectors;
