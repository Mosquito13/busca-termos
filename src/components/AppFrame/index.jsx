import { useEffect, useCallback } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import {
  VscChromeMinimize,
  VscChromeMaximize,
  VscChromeRestore,
  VscChromeClose
} from 'react-icons/vsc';

import { ReactComponent as AppIcon } from '../../assets/svg/icon.svg';
import FrameButton from './FrameButton';

import { close, maximize, minimize, registerFrameListeners, unmaximize } from '../../actions/frame';
import frameSelectors from '../../selectors/frame';

import './styles.scss';

const AppFrame = () => {
  const dispatch = useDispatch();
  const isMaximized = useSelector(frameSelectors.isMaximized);

  const onClose = useCallback(() => dispatch(close()), [dispatch]);
  const onMinimize = useCallback(() => dispatch(minimize()), [dispatch]);
  const onMaximize = useCallback(() => {
    if (isMaximized) {
      return dispatch(unmaximize());
    }
    return dispatch(maximize());
  }, [dispatch, isMaximized]);

  useEffect(() => dispatch(registerFrameListeners()), [dispatch]);

  return (
    <div className="app-frame">
      <div className="app-frame__title">
        <div className="app-frame__title-icon" onDoubleClick={onClose}>
          <AppIcon />
        </div>
      </div>
      <div className="app-frame__buttons">
        <FrameButton
          data-testid="btn-minimize"
          icon={<VscChromeMinimize />}
          onClick={onMinimize}
        />
        <FrameButton
          data-testid="btn-maximize"
          icon={isMaximized ? <VscChromeRestore /> : <VscChromeMaximize />}
          onClick={onMaximize}
        />
        <FrameButton
          close
          data-testid="btn-close"
          icon={<VscChromeClose />}
          onClick={onClose}
        />
      </div>
    </div>
  );
};

export default AppFrame;
