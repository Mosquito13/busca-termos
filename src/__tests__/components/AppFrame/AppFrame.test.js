import { shallow } from 'enzyme';
import renderer from 'react-test-renderer';
import * as reactRedux from 'react-redux';

import AppFrame from '../../../components/AppFrame';
import * as frameActions from '../../../actions/frame';

const getStateMock = (maximized) => ({ Frame: { maximized } });

describe(`${AppFrame.name}`, () => {
  describe('tests', () => {
    let wrapper, dispatchMock, useSelectorSpy, useDispatchSpy;

    beforeAll(() => {
      dispatchMock = jest.fn();

      useSelectorSpy = jest.spyOn(reactRedux, 'useSelector').mockImplementation(selector => selector(getStateMock(false)));
      useDispatchSpy = jest.spyOn(reactRedux, 'useDispatch').mockImplementation(() => dispatchMock);

      jest.spyOn(frameActions, 'close');
      jest.spyOn(frameActions, 'minimize');
      jest.spyOn(frameActions, 'maximize');
      jest.spyOn(frameActions, 'unmaximize');

      wrapper = shallow(
        <AppFrame />
      );
    });

    it('should render unmaximized', () => {
      expect(wrapper.isEmptyRender()).toBeFalsy();
      expect(wrapper.find('.app-frame')).toHaveLength(1);
      expect(wrapper.find('.app-frame__buttons')).toHaveLength(1);
    });

    it('should call close when double click app icon', () => {
      wrapper.find('.app-frame__title-icon').simulate('doubleclick');
      expect(dispatchMock).toBeCalledTimes(1);
      expect(dispatchMock).toBeCalledWith(frameActions.close());
    });

    it('should call minimize', () => {
      wrapper.find('[data-testid="btn-minimize"]').simulate('click');
      expect(dispatchMock).toBeCalledTimes(1);
      expect(dispatchMock).toBeCalledWith(frameActions.minimize());
    });

    it('should call maximize', () => {
      wrapper.find('[data-testid="btn-maximize"]').simulate('click');
      expect(dispatchMock).toBeCalledTimes(1);
      expect(dispatchMock).toBeCalledWith(frameActions.maximize());
    });

    it('should render maximized', () => {
      wrapper.unmount();
      useSelectorSpy.mockImplementation(selector => selector(getStateMock(true)));
      useDispatchSpy.mockImplementation(() => dispatchMock);

      wrapper = shallow(
        <AppFrame />
      );

      expect(wrapper.find('.app-frame')).toHaveLength(1);
      expect(wrapper.find('.app-frame__buttons')).toHaveLength(1);
    });

    it('should call unmaximize', () => {
      wrapper.find('[data-testid="btn-maximize"]').simulate('click');
      expect(dispatchMock).toBeCalledTimes(1);
      expect(dispatchMock).toBeCalledWith(frameActions.unmaximize());
    });

    afterAll(() => {
      wrapper.unmount();
    });
  });

  describe('snapshots', () => {
    let useSelectorSpy;

    beforeEach(() => {
      jest.spyOn(reactRedux, 'useDispatch').mockImplementation(() => jest.fn());
      useSelectorSpy = jest.spyOn(reactRedux, 'useSelector');
    });

    it('should render maximized', () => {
      useSelectorSpy.mockImplementation(selector => selector(getStateMock(true)));

      const tree = renderer
        .create(<AppFrame />)
        .toJSON();

      expect(tree).toMatchSnapshot();
    });

    it('should render unmaximized', () => {
      useSelectorSpy.mockImplementation(selector => selector(getStateMock(false)));

      const tree = renderer
        .create(<AppFrame />)
        .toJSON();

      expect(tree).toMatchSnapshot();
    });
  });
});
