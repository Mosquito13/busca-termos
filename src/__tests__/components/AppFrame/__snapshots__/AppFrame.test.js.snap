// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppFrame snapshots should render maximized 1`] = `
<div
  className="app-frame"
>
  <div
    className="app-frame__title"
  >
    <div
      className="app-frame__title-icon"
      onDoubleClick={[Function]}
    >
      <svg>
        icon.svg
      </svg>
    </div>
  </div>
  <div
    className="app-frame__buttons"
  >
    <button
      className="app-frame__button"
      onClick={[Function]}
    >
      <div
        className="app-frame__button-icon"
      >
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          strokeWidth="0"
          style={
            Object {
              "color": undefined,
            }
          }
          viewBox="0 0 16 16"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M14 8v1H3V8h11z"
          />
        </svg>
      </div>
    </button>
    <button
      className="app-frame__button"
      onClick={[Function]}
    >
      <div
        className="app-frame__button-icon"
      >
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          strokeWidth="0"
          style={
            Object {
              "color": undefined,
            }
          }
          viewBox="0 0 16 16"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M3 5v9h9V5H3zm8 8H4V6h7v7z"
          />
          <path
            clipRule="evenodd"
            d="M5 5h1V4h7v7h-1v1h2V3H5v2z"
            fillRule="evenodd"
          />
        </svg>
      </div>
    </button>
    <button
      className="app-frame__button app-frame__button--close"
      onClick={[Function]}
    >
      <div
        className="app-frame__button-icon"
      >
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          strokeWidth="0"
          style={
            Object {
              "color": undefined,
            }
          }
          viewBox="0 0 16 16"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            clipRule="evenodd"
            d="M7.116 8l-4.558 4.558.884.884L8 8.884l4.558 4.558.884-.884L8.884 8l4.558-4.558-.884-.884L8 7.116 3.442 2.558l-.884.884L7.116 8z"
            fillRule="evenodd"
          />
        </svg>
      </div>
    </button>
  </div>
</div>
`;

exports[`AppFrame snapshots should render unmaximized 1`] = `
<div
  className="app-frame"
>
  <div
    className="app-frame__title"
  >
    <div
      className="app-frame__title-icon"
      onDoubleClick={[Function]}
    >
      <svg>
        icon.svg
      </svg>
    </div>
  </div>
  <div
    className="app-frame__buttons"
  >
    <button
      className="app-frame__button"
      onClick={[Function]}
    >
      <div
        className="app-frame__button-icon"
      >
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          strokeWidth="0"
          style={
            Object {
              "color": undefined,
            }
          }
          viewBox="0 0 16 16"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M14 8v1H3V8h11z"
          />
        </svg>
      </div>
    </button>
    <button
      className="app-frame__button"
      onClick={[Function]}
    >
      <div
        className="app-frame__button-icon"
      >
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          strokeWidth="0"
          style={
            Object {
              "color": undefined,
            }
          }
          viewBox="0 0 16 16"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M3 3v10h10V3H3zm9 9H4V4h8v8z"
          />
        </svg>
      </div>
    </button>
    <button
      className="app-frame__button app-frame__button--close"
      onClick={[Function]}
    >
      <div
        className="app-frame__button-icon"
      >
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          strokeWidth="0"
          style={
            Object {
              "color": undefined,
            }
          }
          viewBox="0 0 16 16"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            clipRule="evenodd"
            d="M7.116 8l-4.558 4.558.884.884L8 8.884l4.558 4.558.884-.884L8.884 8l4.558-4.558-.884-.884L8 7.116 3.442 2.558l-.884.884L7.116 8z"
            fillRule="evenodd"
          />
        </svg>
      </div>
    </button>
  </div>
</div>
`;
